src/Rules/Deprecations/EchoDeprecatedToStringRule.php

```diff
@@ -0,0 +1,115 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Deprecations;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Echo_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\RuleLevelHelper;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+/**
+ * @implements \PHPStan\Rules\Rule<Echo_>
+ */
+class EchoDeprecatedToStringRule implements \PHPStan\Rules\Rule
+{
+
+	/** @var RuleLevelHelper */
+	private $ruleLevelHelper;
+
+	public function __construct(RuleLevelHelper $ruleLevelHelper)
+	{
+		$this->ruleLevelHelper = $ruleLevelHelper;
+	}
+
+	public function getNodeType(): string
+	{
+		return Echo_::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (DeprecatedScopeHelper::isScopeDeprecated($scope)) {
+			return [];
+		}
+
+		$messages = [];
+		foreach ($node->exprs as $key => $expr) {
+			$this->deepCheckExpr($expr, $scope, $messages);
+		}
+
+		return $messages;
+	}
+
+	/**
+	 * @param Node\Expr $expr
+	 * @param Scope $scope
+	 * @param string[] $messages
+	 */
+	private function deepCheckExpr(Node\Expr $expr, Scope $scope, array &$messages): void
+	{
+		if ($expr instanceof Node\Expr\BinaryOp) {
+			$this->deepCheckExpr($expr->left, $scope, $messages);
+			$this->deepCheckExpr($expr->right, $scope, $messages);
+		} elseif ($expr instanceof Node\Expr\Cast\String_ || $expr instanceof Node\Expr\Assign || $expr instanceof Node\Expr\AssignOp) {
+			$this->deepCheckExpr($expr->expr, $scope, $messages);
+		} elseif ($expr instanceof Node\Scalar\Encapsed) {
```

Comment:
All of these should have separate rules, exactly as it's in phpstan/phpstan. You don't have to do "deep check" here because the nested expressions like Concat will get picked up by the other rules.